pmd-core/src/main/java/net/sourceforge/pmd/lang/metrics/ParameterizedMetricKey.java

```diff
@@ -57,13 +59,22 @@ private static int code(MetricKey key, MetricVersion version) {
     }
 
 
-    /** Builds a parameterized metric key. */
-    public static ParameterizedMetricKey getInstance(MetricKey<? extends Node> key, MetricVersion version) {
+    /**
+     * Builds a parameterized metric key.
+     *
+     * @param key     The key
+     * @param version The version
+     * @param <N>     The type of node of the metrickey
+     *
+     * @return An instance of parameterized metric key corresponding to the parameters
+     */
+    @SuppressWarnings("unchecked")
+    public static <N extends Node> ParameterizedMetricKey<N> getInstance(MetricKey<N> key, MetricVersion version) {
         int code = code(key, version);
```

Comment:
You may be right, we could as well rely on HashMap's behaviour. Plus that makes one less method in the class, which is nice. We're not pooling them for performance anyway, but mostly to keep the number of instances in memory small. Calculating 10 class metrics on a 100 classes project would keep 1000 keys in memory without pooling, but only 10 with pooling. That might help if we serialise the project mirror to cache results